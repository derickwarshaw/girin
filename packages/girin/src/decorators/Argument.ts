import { ArgumentMetadata } from '../metadata/ArgumentMetadata';
import { Generic } from '../metadata/Generic';


export interface ArgumentDecoratorOptions {
  defaultValue?: any;
  description?: string;
}

export function Argument(nameAndType: string, options?: ArgumentDecoratorOptions) {
  const [name, typeString] = nameAndType.split(/\s*:\s*/);

  return function(target: Object, fieldName: string, parameterIndex: number) {
    const definitionClass = target.constructor;
    ArgumentMetadata.create({
      definitionClass,
      fieldName,
      name,
      definedOrder: parameterIndex,
      generic: Generic.of(typeString),
      defaultValue: options && options.defaultValue,
      description: options && options.description,
    });
  }
}