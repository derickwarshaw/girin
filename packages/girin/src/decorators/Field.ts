import { FieldMetadata } from '../metadata/FieldMetadata';
import { Generic } from "../metadata/Generic";
import { globalMetadataStorage } from "../metadata/globalMetadataStorage";

export interface FieldDecoratorOptions {
  description?: string;
  deprecationReason?: string;
}

export function Field(type: string, options?: FieldDecoratorOptions): Function {
  return function(objectType: any, fieldName: string) {
    const definitionClass = objectType.constructor as any;
    FieldMetadata.create({
      fieldName,
      definitionClass,
      description: options && options.description,
      deprecationReason: options && options.deprecationReason,
      generic: Generic.of(type),
      meta: globalMetadataStorage,
    });
  };
}
