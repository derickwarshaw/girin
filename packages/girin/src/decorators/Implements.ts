import { MetadataStorage } from '../metadata/MetadataStorage';
import { ImplementsMetadata } from '../metadata/ImplementsMetadata';


export interface ImplementsDecoratorOptions {
  meta?: MetadataStorage;
}

export function Implements(targetNameOrDefinitionClass: Function | string, options: ImplementsDecoratorOptions = {}) {
  return function(definitionClass: Function & ImplementsDecoratorOptions) {
    const mergedConfig = Object.assign(options, definitionClass, { definitionClass, targetNameOrDefinitionClass });
    ImplementsMetadata.create(mergedConfig);
  };
}