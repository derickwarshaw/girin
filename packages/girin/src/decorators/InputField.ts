
import { InputFieldMetadata } from "../metadata/InputFieldMetadata";
import { MetadataStorage } from "../metadata/MetadataStorage";
import { Generic } from "../metadata/Generic";


export interface InputFieldDecoratorOptions {
  defaultValue?: any;
  description?: string;
  meta?: MetadataStorage;
}

export function InputField(nameAndType: string, options?: InputFieldDecoratorOptions): Function {
  const [ name, typeString ] = nameAndType.split(/\s*:\s*/);

  return function(definitionClass: Function, propertyName: undefined, parameterIndex: number) {
    InputFieldMetadata.create({
      definedOrder: parameterIndex,
      fieldName: name,
      generic: Generic.of(typeString),
      definitionClass: definitionClass,
      description: options && options.description,
      defaultValue: options && options.defaultValue,
      meta: options && options.meta,
    });
  };
}

