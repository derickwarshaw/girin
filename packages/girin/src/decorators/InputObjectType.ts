import { GraphQLInputFieldConfigMap, GraphQLInputObjectTypeConfig } from 'graphql';

import { MetadataStorage } from '../metadata/MetadataStorage';
import { InputObjectTypeMetadata } from '../metadata/InputObjectTypeMetadata';


export type InputObjectTypeDecoratorOptions = {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLInputFieldConfigMap;
  description?: string;
  astNode?: GraphQLInputObjectTypeConfig["astNode"];
}

export function InputObjectType(options: InputObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: InputObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    InputObjectTypeMetadata.create(mergedConfig);
  };
}
