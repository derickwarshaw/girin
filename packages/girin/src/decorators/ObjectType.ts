import { ObjectTypeMetadata } from '../metadata/ObjectTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLObjectTypeConfig, GraphQLFieldConfigMap } from 'graphql';


export interface ObjectTypeDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  interfaces?: GraphQLObjectTypeConfig<any, any>["interfaces"];
  isTypeOf?: GraphQLObjectTypeConfig<any, any>["isTypeOf"];
  description?: GraphQLObjectTypeConfig<any, any>["description"];
  astNode?: GraphQLObjectTypeConfig<any, any>["astNode"];
  extensionASTNodes?: GraphQLObjectTypeConfig<any, any>["extensionASTNodes"];
}

export function ObjectType(options: ObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: ObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    ObjectTypeMetadata.create(mergedConfig);
  };
}
