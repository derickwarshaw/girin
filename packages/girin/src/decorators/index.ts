export { Argument } from './Argument';
export { Field } from './Field';
export { Implements } from './Implements';
export { InputField } from './InputField';
export { InputObjectType } from './InputObjectType';
export { ObjectType } from './ObjectType';
