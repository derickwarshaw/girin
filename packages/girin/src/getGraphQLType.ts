import { MetadataStorage } from "./metadata/MetadataStorage";
import { GraphQLType, isType } from 'graphql';

export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(definitionClassOrType: Function | GraphQLType, options?: TypeBuilderOptions): any {
  let meta: MetadataStorage;
  if (options && options.meta) {
    meta = options.meta;
  } else {
    meta = MetadataStorage.getMetadataStorage();
  }

  if (isType(definitionClassOrType)) {
    return definitionClassOrType;
  }

  const typeMetadata = meta.getDefinitionMetadata(definitionClassOrType);
  if (!typeMetadata) {
    throw new Error(`Given class ${definitionClassOrType.name} has no corresponding metadata`);
  }
  return typeMetadata.build.typeInstance;
}
