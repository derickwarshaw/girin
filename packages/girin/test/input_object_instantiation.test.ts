import { graphql, GraphQLSchema, printSchema } from 'graphql';

import { Field } from '../src/decorators/Field';
import { Argument } from '../src/decorators/Argument';
import { InputField } from '../src/decorators/InputField';
import { ObjectType } from '../src/decorators/ObjectType';
import { InputObjectType } from '../src/decorators/InputObjectType';
import { getGraphQLType } from '../src/getGraphQLType';


interface MemberSource {
  id: number;
  username: string;
  email: string;
}

@ObjectType()
class Member {
  constructor(source: MemberSource) {
    this.id = source.id;
    this.username = source.username;
    this.email = source.email;
  }
  @Field('Int!') id: number;
  @Field('String!') username: string;
  @Field('String!') email: string;
}

@InputObjectType()
class MemberProfileInput {
  constructor(
    @InputField("firstName: String!")
    public firstName: string,
    @InputField("lastName: String!")
    public lastName: string,
  ) { }
}

@InputObjectType()
class MemberInput {
  constructor(
    @InputField("username: String!")
    public username: string,
    @InputField("email: String!")
    public email: string,
    @InputField('profile: MemberProfileInput!')
    public profile: MemberProfileInput,
  ) { }
}

@ObjectType()
class Query {
  @Field('Member!')
  public getMember() {
    return new Member({ id: 1, username: 'Jonghyun', email: 'j@example.com' });
  }
}

@ObjectType()
class Mutation {
  @Field('Boolean!')
  public instantiateMemberInput(
    @Argument("member: MemberInput!") member: MemberInput,
  ) {
    return member instanceof MemberInput && member.profile instanceof MemberProfileInput;
  }

  @Field('String!')
  public getEmailFromInput(
    @Argument("member: MemberInput!") member: MemberInput,
  ) {
    return member.email;
  }

  @Field('String!')
  public getUsernameFromInput(
    @Argument("member: MemberInput!") member: MemberInput,
  ) {
    return member.username;
  }
}

const schema = new GraphQLSchema({
  query: getGraphQLType(Query),
  mutation: getGraphQLType(Mutation),
});


describe('Input object instantiation', () => {
  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('should create InputObjectType instance with given argument value', async () => {
    const result = await graphql({ schema, source: `
      mutation {
        instantiateMemberInput(member: {
          username: "key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
        getUsernameFromInput(member: {
          username: "Key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
        getEmailFromInput(member: {
          username: "Key",
          email: "k@example.com",
          profile: { firstName: "Kibum", lastName: "Kim" }
        })
      }
    `});
    // console.log(JSON.stringify(result, null, '  '));
    expect(result.data!.getUsernameFromInput).toBe("Key");
    expect(result.data!.getEmailFromInput).toBe("k@example.com");
    expect(result.data!.instantiateMemberInput).toBe(true);
  });
});
