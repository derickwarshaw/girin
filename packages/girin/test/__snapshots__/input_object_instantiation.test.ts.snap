// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input object instantiation generates schema as expected 1`] = `
"type Member {
  id: Int!
  username: String!
  email: String!
}

input MemberInput {
  profile: MemberProfileInput!
  email: String!
  username: String!
}

input MemberProfileInput {
  lastName: String!
  firstName: String!
}

type Mutation {
  instantiateMemberInput(member: MemberInput!): Boolean!
  getEmailFromInput(member: MemberInput!): String!
  getUsernameFromInput(member: MemberInput!): String!
}

type Query {
  getMember: Member!
}
"
`;
